libservice/src/Aggregator.cpp

```diff
@@ -28,60 +28,101 @@ static std::string getEndpoint(const std::string& host, const std::string& url)
-	bool isPossiblyIpv6{std::count(addr.begin(), addr.end(), ':') >= 2};
-	return isClientExternal(ipChecker, addr, isPossiblyIpv6);
+static bool isClientExternal(const IpAddressChecker& ipChecker, const std::variant<in_addr, in6_addr>& clientAddrBinary, bool isIpv6) {
+	return isIpv6 ? isIpv6ClientExternal(ipChecker, clientAddrBinary) : isIpv4ClientExternal(ipChecker, clientAddrBinary);
 }
 
+static bool enableNetworkCounters;
 static void incrementServiceClientsNumber(
-		const IpAddressChecker& ipChecker, Service& service, const httpparser::HttpRequest& request, const DiscoverySessionMeta& meta) {
+		const IpAddressChecker& ipChecker, Service& service, const httpparser::HttpRequest& request, const DiscoverySessionMeta& meta, const std::chrono::time_point<std::chrono::steady_clock>& currentTime) {
 	bool isExternal{false};
+	bool isIpv6{false};
 	std::string clientAddr;
+
+	std::variant<in_addr, in6_addr> clientAddrBinary{};
 	try {
 		if (!request.clientIp.empty()) {
 			clientAddr = request.clientIp.front();
-			isExternal = isClientExternal(ipChecker, clientAddr);
+			if (std::count(clientAddr.begin(), clientAddr.end(), ':') >= 2) { //Is possibly IPv6?
+				isIpv6 = true;
+			}
 		} else if (discoveryFlagsSessionIsIPv4(meta.flags)) {
 			clientAddr = ipv4ToString(meta.sourceIP.data);
-			isExternal = isClientExternal(ipChecker, clientAddr, false);
 		} else if (discoveryFlagsSessionIsIPv6(meta.flags)) {
 			clientAddr = ipv6ToString(meta.sourceIP.data);
-			isExternal = isClientExternal(ipChecker, clientAddr, true);
+			isIpv6 = true;
 		} else {
 			return;
 		}
+
+		if (isIpv6) {
+			clientAddrBinary = in6_addr{};
+			if (inet_pton(AF_INET6, clientAddr.c_str(), &clientAddrBinary) != 1) {
+				throw std::runtime_error("Couldn't parse IPv6 client address");
+			}
+		} else {
+			clientAddrBinary = in_addr{};
+			if (inet_pton(AF_INET, clientAddr.c_str(), &std::get<in_addr>(clientAddrBinary)) != 1) {
+				throw std::runtime_error("Couldn't parse IPv4 client address");
+			}
+		}
+		isExternal = isClientExternal(ipChecker, clientAddrBinary, isIpv6);
 	} catch (const std::runtime_error& e) {
 		LOG_TRACE("Couldn't determine if the client is external: {} (client address: {})", e.what(), clientAddr);
+		return;
+	} catch (const std::bad_variant_access& e) {
+		LOG_TRACE("Bad variant access: {} (client address: {})", e.what(), clientAddr);
+		return;
 	}
 
 	if (isExternal) {
 		++service.externalClientsNumber;
+
+		if (enableNetworkCounters) {
+			try {
+				if (isIpv6) {
+					std::array<uint8_t, service::ipv6NetworkPrefixBytesLen> networkIPv6{};
+					std::memcpy(networkIPv6.data(), std::get<in6_addr>(clientAddrBinary).s6_addr, service::ipv6NetworkPrefixBytesLen);
+
+					if (auto it = service.detectedExternalIPv6Networks.find(networkIPv6); it != service.detectedExternalIPv6Networks.end()) {
+						it->second = currentTime;
+					} else {
+						service.detectedExternalIPv6Networks[networkIPv6] = currentTime;
```

Comment:
Why isn't it used for both cases? (if-else and find in map seems to be redundant)